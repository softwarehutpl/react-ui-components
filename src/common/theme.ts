const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
