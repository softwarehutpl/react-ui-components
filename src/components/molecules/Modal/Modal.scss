.modal {
    width: 100vw;
    height: 100vh;
    background-color: black;
    opacity: 0.8;
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    top: 0;
}