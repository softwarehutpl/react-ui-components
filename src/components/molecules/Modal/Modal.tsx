import React from 'react';
import styled from 'styled-components';
import ReactDOM from 'react-dom';
import CloseIcon from '../../../common/icons/CloseIcon/CloseIcon';
import './Modal.scss';

interface ModalProps {
  onClose: () => void;
  rootId: string;
  isOpen: boolean;
  topButtonPosition?: number;
  rightButtonPosition?: number;
  buttonFontColor?: string;
  buttonBackgroundColor?: string;
  showCloseButton?: true;
  backgroundColor?: string;
  opacity?: number;
  closeButtonText?: string;
  children?: any;
  showTransitionEffect?: boolean;
  transitionEffect?: string;
  ownCloseButtonIcon?: any;
  classicCloseButton?: boolean;
  fontSize?: number;
  contentWidth: number;
  contentHeight: number;
}

interface ModalWrapperProps {
  isOpen: boolean;
  backgroundColor?: string;
  showTransitionEffect?: boolean;
  transitionEffect?: string;
  opacity?: number;
}

export const ModalWrapper = styled.div<ModalWrapperProps>`
  width: 100vw;
  height: 100vh;
  background-color: ${({ backgroundColor, isOpen }) => isOpen ? backgroundColor : 'transparent'};
  opacity: ${({ opacity }) => opacity};
  position: fixed;
  display: flex;
  visibility: ${({ isOpen }) => isOpen ? 'visible' : 'hidden'};
  transition: ${({ showTransitionEffect, transitionEffect, theme }) => showTransitionEffect ? (transitionEffect ? theme.transitions.all[transitionEffect] : 'all 1s') : 'none'};
  justify-content: center;
  align-items: center;
  top: 0;
  z-index: 1;
`;

export const ModalContent = styled.div<any>`
  position: absolute;
  background-color: ${({ contentBackgroundColor, isOpen }) => isOpen ? contentBackgroundColor : 'transparent'};
  width: ${({ contentWidth }) => `${contentWidth}px`};
  height: ${({ contentHeight }) => `${contentHeight}px`};
  visibility: ${({ isOpen }) => isOpen ? 'visible' : 'hidden'};
  color: ${({ isOpen }) => isOpen ? 'inherit' : 'transparent'};
  transition: all 1s ease;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const ModalToggleButton = styled.button<any>`
  position: absolute;
  top: ${({ closeButtonOutside, topButtonPosition }) => !closeButtonOutside ? `${topButtonPosition}px` : 0};
  right: ${({ closeButtonOutside, rightButtonPosition }) => !closeButtonOutside ? `${rightButtonPosition}px` : 0};
  color: ${({ buttonFontColor }) => buttonFontColor};
  transform: ${({ closeButtonOutside, topButtonPosition, rightButtonPosition, buttonWidth }) => closeButtonOutside && `translate(${rightButtonPosition + buttonWidth}px, -${topButtonPosition}px)`};
  background-color: ${({ buttonBackgroundColor }) => buttonBackgroundColor};
  z-index: 1;
  border: none;
  cursor: pointer;
  width: ${({ buttonWidth }) => `${buttonWidth}px`};
  text-align: ${({ closeButtonOutside }) => closeButtonOutside ? 'left' : 'right'};
`;

const ModalCloseIcon = styled.div<any>`
  position: absolute;
  top: ${({ closeButtonOutside, topButtonPosition }) => !closeButtonOutside ? `${topButtonPosition}px` : 0};
  right: ${({ closeButtonOutside, rightButtonPosition }) => !closeButtonOutside ? `${rightButtonPosition}px` : 0};
  transform: ${({ closeButtonOutside, rightButtonPosition, fontSize }) => closeButtonOutside && `translate(${rightButtonPosition}px, -${fontSize}px)`};
  color: ${({ buttonFontColor }) => buttonFontColor};
  background-color: ${({ buttonBackgroundColor }) => buttonBackgroundColor};
  z-index: 1;
  border: none;
  cursor: pointer;
`;

const Modal = (props: ModalProps) => {
 
  const modalRoot = typeof document !== 'undefined' ? document.getElementById(props.rootId) : null;

  if(!modalRoot) {
    return null;
  }

  return ReactDOM.createPortal(
    <ModalWrapper
      id="modal"
      isOpen={props.isOpen}
      backgroundColor={props.backgroundColor}
      showTransitionEffect={props.showTransitionEffect}
      opacity={props.opacity}
    >
      <ModalContent {...props}>{props.children}
      {props.showCloseButton && !props.ownCloseButtonIcon && !props.classicCloseButton &&
        <ModalToggleButton {...props} onClick={props.onClose}>
          {props.closeButtonText}
        </ModalToggleButton>}
      {props.ownCloseButtonIcon && !props.classicCloseButton && <ModalCloseIcon {...props}>
        {props.ownCloseButtonIcon}
      </ModalCloseIcon>}
        {props.classicCloseButton && <ModalCloseIcon {...props}>
          <CloseIcon
            onClick={props.onClose}
            iconColor={props.isOpen ? props.buttonFontColor : 'transparent'}
            topPosition={props.topButtonPosition}
            rightPosition={props.rightButtonPosition}
            height={`${props.fontSize}px`}
          />
        </ModalCloseIcon>}
      </ModalContent>
    </ModalWrapper>,
    modalRoot,
  );
}

const defaultProps = {
  topButtonPosition: 0,
  rightButtonPosition: 0,
  buttonFontColor: '#fff',
  buttonBackgroundColor: 'transparent',
  showCloseButton: true,
  backgroundColor: 'black',
  opacity: 0.6,
  closeButtonText: 'CLOSE',
  showTransitionEffect: false,
  closeButtonOutside: false,
  buttonWidth: 120,
  contentWidth: 300,
  contentHeight: 300,
  contentBackgroundColor: '#fff',
  fontSize: 24,
};

Modal.defaultProps = defaultProps;

export default Modal;
