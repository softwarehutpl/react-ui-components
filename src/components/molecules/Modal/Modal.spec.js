
import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Modal, { ModalToggleButton, ModalWrapper } from './Modal';
import theme from '../../../common/theme';

describe('Modal Component', () => {
  it('should render correctly', () => {
    const tree = create(<Modal onClose={() => console.log()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call  modal toggle component with onClose method', () => {
    const mockOnClick = jest.fn();
    const component = create(<ModalToggleButton onClose={mockOnClick} />);
    const instance = component.root;
    const button = instance.findByType('button');
    act(() => {
      button.props.onClose();
    });
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });

  it('should render Modal component introduced styles props', () => {
    const component = create(<ModalWrapper isOpen backgroundColor='hotpink' opacity={0.7} />);
    const modalWrapper = component.toJSON();
    expect(modalWrapper).toHaveStyleRule('background-color', 'hotpink');
  });

  it('should render Modal component with visibbility hidden when isOpen is false', () => {
    const component = create(<ModalWrapper isOpen={false} />);
    const modalWrapper = component.toJSON();
    expect(modalWrapper).toHaveStyleRule('visibility', 'hidden');
  });

  it('should render Modal component with visibbility visible when isOpen is true', () => {
    const component = create(<ModalWrapper isOpen />);
    const modalWrapper = component.toJSON();
    expect(modalWrapper).toHaveStyleRule('visibility', 'visible');
  });

  it('should render modal toggle button when modal is open', () => {
    const component = create(<ModalToggleButton
      theme={theme}
      buttonFontColor='grey'
      buttonBackgroundColor='hotpink'
      topButtonPosition={30}
      rightButtonPosition={30}
      isOpen
    />)
    const input = component.toJSON();
    expect(input).toHaveStyleRule('color', 'grey');
    expect(input).toHaveStyleRule('background-color', 'hotpink');
    expect(input).toHaveStyleRule('top', '30px');
    expect(input).toHaveStyleRule('right', '30px');
  });
});
