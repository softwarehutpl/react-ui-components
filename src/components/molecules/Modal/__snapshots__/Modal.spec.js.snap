// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should render correctly 1`] = `null`;
